File: kurrent_skills/templates/python/persistent_subscription_example.py
#!/usr/bin/env python3
"""KurrentDB Python Client Example - Persistent Subscriptions with all NACK actions"""

import json
import uuid
from kurrentdbclient import KurrentDBClient, NewEvent, StreamState

# === CONNECTION ===
client = KurrentDBClient(uri="kurrentdb://admin:changeit@localhost:2113?tls=false")
print("Connected to KurrentDB")

stream_name = "orders"
group_name = "order-processor"

# === NACK ACTIONS ===
# Available NACK actions for persistent subscriptions:
# - "park"  : Move to parked queue for manual inspection
# - "retry" : Retry processing immediately
# - "skip"  : Skip this event and continue
# - "stop"  : Stop the subscription

# === CREATE PERSISTENT SUBSCRIPTION ===
try:
    client.create_subscription_to_stream(
        group_name=group_name,
        stream_name=stream_name
    )
    print(f"Created persistent subscription '{group_name}' on stream '{stream_name}'")
except Exception as e:
    if "already exists" in str(e).lower():
        print(f"Persistent subscription '{group_name}' already exists")
    else:
        raise

# === APPEND SOME TEST EVENTS ===
for i in range(3):
    event = NewEvent(
        type="OrderCreated",
        data=json.dumps({
            "order_id": str(uuid.uuid4()),
            "amount": 10.00 * (i + 1)
        }).encode("utf-8"),
        content_type="application/json"
    )
    client.append_to_stream(
        stream_name=stream_name,
        current_version=StreamState.ANY,
        events=[event]
    )
    print(f"Appended OrderCreated event #{i + 1}")

# === SUBSCRIBE TO PERSISTENT SUBSCRIPTION ===
print("\nSubscribing to persistent subscription...")

subscription = client.read_subscription_to_stream(
    group_name=group_name,
    stream_name=stream_name
)

count = 0
for event in subscription:
    try:
        print(f"  Processing: {event.type}")
        print(f"  Data: {event.data.decode('utf-8')}")

        # Simulate processing with different outcomes
        data = json.loads(event.data.decode('utf-8'))
        amount = data.get("amount", 0)

        if amount > 25:
            # Simulate transient failure - retry immediately
            subscription.nack(event.ack_id, action="retry")
            print(f"  Retrying event (transient failure)")
        elif amount > 20:
            # Simulate permanent failure - park for inspection
            subscription.nack(event.ack_id, action="park")
            print(f"  Parked event (permanent failure)")
        elif amount > 15:
            # Skip malformed/invalid event
            subscription.nack(event.ack_id, action="skip")
            print(f"  Skipped event (invalid data)")
        else:
            # Success - acknowledge
            subscription.ack(event.ack_id)
            print(f"  Acknowledged event")

    except Exception as e:
        # Park failed event for later inspection
        subscription.nack(event.ack_id, action="park")
        print(f"  Parked event due to error: {e}")

    count += 1
    if count >= 3:
        break

# Note: action="stop" would stop the subscription entirely
# subscription.nack(event.ack_id, action="stop")

print("\nDone!")

File: kurrent_skills/templates/nodejs/persistent_subscription_example.js
/**
 * KurrentDB Node.js Client Example - Persistent Subscriptions with all NACK actions
 */

import {
    KurrentDBClient,
    jsonEvent,
    persistentSubscriptionToStreamSettingsFromDefaults,
    PARK,
    RETRY,
    SKIP,
    STOP,
} from "@kurrent/kurrentdb-client";
import { v4 as uuid } from "uuid";

// === NACK ACTIONS ===
// Available NACK actions for persistent subscriptions:
// - PARK  : Move to parked queue for manual inspection
// - RETRY : Retry processing immediately
// - SKIP  : Skip this event and continue
// - STOP  : Stop the subscription

// === CONNECTION ===
const client = KurrentDBClient.connectionString`kurrentdb://admin:changeit@localhost:2113?tls=false`;
console.log("Connected to KurrentDB");

const streamName = "orders";
const groupName = "order-processor";

// === CREATE PERSISTENT SUBSCRIPTION ===
try {
    await client.createPersistentSubscriptionToStream(
        streamName,
        groupName,
        persistentSubscriptionToStreamSettingsFromDefaults()
    );
    console.log(`Created persistent subscription '${groupName}' on stream '${streamName}'`);
} catch (error) {
    if (error.message?.includes("already exists")) {
        console.log(`Persistent subscription '${groupName}' already exists`);
    } else {
        throw error;
    }
}

// === APPEND SOME TEST EVENTS ===
for (let i = 0; i < 3; i++) {
    const event = jsonEvent({
        type: "OrderCreated",
        data: {
            orderId: uuid(),
            amount: 10.0 * (i + 1),
        },
    });
    await client.appendToStream(streamName, event);
    console.log(`Appended OrderCreated event #${i + 1}`);
}

// === SUBSCRIBE TO PERSISTENT SUBSCRIPTION ===
console.log("\nSubscribing to persistent subscription...");

const subscription = client.subscribeToPersistentSubscriptionToStream(
    streamName,
    groupName
);

let count = 0;
try {
    for await (const event of subscription) {
        try {
            console.log(`  Processing: ${event.event?.type} (retry: ${event.retryCount})`);
            console.log(`  Data: ${JSON.stringify(event.event?.data)}`);

            // Simulate processing with different outcomes based on amount
            const amount = event.event?.data?.amount || 0;

            if (amount > 25) {
                // Simulate transient failure - retry immediately
                await subscription.nack(RETRY, "Transient failure - retrying", event);
                console.log(`  Retrying event (transient failure)`);
            } else if (amount > 20) {
                // Simulate permanent failure - park for inspection
                await subscription.nack(PARK, "Permanent failure - parking", event);
                console.log(`  Parked event (permanent failure)`);
            } else if (amount > 15) {
                // Skip malformed/invalid event
                await subscription.nack(SKIP, "Invalid data - skipping", event);
                console.log(`  Skipped event (invalid data)`);
            } else {
                // Success - acknowledge
                await subscription.ack(event);
                console.log(`  Acknowledged event`);
            }
        } catch (error) {
            // Park failed event for later inspection
            await subscription.nack(PARK, error.toString(), event);
            console.log(`  Parked event due to error: ${error.message}`);
        }

        count++;
        if (count >= 3) {
            break;
        }
    }
} catch (error) {
    console.log(`Subscription was dropped: ${error}`);
}

// Note: STOP would stop the subscription entirely
// await subscription.nack(STOP, "Stopping subscription", event);

console.log("\nDone!");
process.exit(0);
